docs/proposals/volume-provisioning.md

```diff
@@ -121,30 +124,125 @@ We propose that:
+
+  3. If tries to find a StorageClass instance referenced by annotation
+     `volume.beta.kubernetes.io/storage-class`. If not found, it reports
+     an error and goes back to step i. periodically.
+
+  4. The provisioner parses arguments in the `StorageClass` and `claim.Labels`
+     and provisions appropriate storage asset. On error, it sends all errors as
+     events on the claim and goes back to step i., i.e. re-tries after a while.
+
+  5. When the volume is provisioned, it submits a new PV representing the
+     storage asset. When this fails, it SHOULD retry creating the PV again few
+     times. If all attempts fail, it MUST delete the storage asset. All errors
+     are sent as events to the claim.
+
+     The created PV MUST have these properties:
+
+     * `pv.Spec.ClaimRef` must point to the claim that led to its creation
+     (including the claim UID).
+
+     * `pv.Annotations["pv.kubernetes.io/provisioned-by"]` must be set to name
+       of the external provisioner. This provisioner will be used to delete the
+       volume!
+
+     * The provisioner can store any other information it needs to the created
+       PV as annotations.
+
+  As result, Kubernetes has a PV that represents the storage asset and is bound to
+  the claim. From now on, usual processing happens, i.e. the volume controller
+  either binds the PV to the claim or deletes the storage asset (see below for
+  external deleters) when the claim was deleted or bound to another PV in the
+  meantime.
+
+### Controller workflow for deleting volumes
+
+When the controller decides that a volume should be deleted it performs these
+steps:
 
-5.  All provisioners are in-tree; they implement an interface called
-    `ProvisionableVolumePlugin`, which has a method called `NewProvisioner`
-    that returns a new provisioner.
+1. The controller changes `pv.Status.Phase` to `Released`.
 
-6.  The controller calls volume plugin `Provision` with ProvisionerParameters from the `StorageClass` configuration object.
+2. The controller looks for `pv.Annotations["pv.kubernetes.io/provisioned-by"]`.
+   If found, it uses this provisioner/deleter to delete the volume.
 
-7.  If `Provision` returns an error, the controller generates an event on the
-    claim and goes back to step 1., i.e. it will retry provisioning periodically
+3. If the volume is not annotated by `pv.kubernetes.io/provisioned-by`, the
+   controller inspects `pv.Spec` and finds in-tree deleter for the volume.
 
-8.  If `Provision` returns no error, the controller creates the returned
-    `api.PersistentVolume`, fills its `Class` attribute with `claim.Spec.Class`
-    and makes it already bound to the claim
+4. If the deleter found by steps 2. or 3. is internal, it calls it and deletes
+   the storage asset together with the PV that represents it.
 
-  1.  If the create operation for the `api.PersistentVolume` fails, it is
-      retried
+5. If the deleter is not known to Kubernetes, it does not do anything.
```

Comment:
I thought the new annotation was the only change needed to implement this proposal, this is another one since right now kubernetes sets PV to failed if it can't find a deleter. 

Just noting because I know I said annotation's the only change a couple times